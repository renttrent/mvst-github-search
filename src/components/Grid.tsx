import styled from "@emotion/styled"

const GridCSS = styled('div')`
  @media (min-width: 0px) {
    height: 55vh;
  }
  @media (min-width: 1300px) {
    height: 63vh;
  }
  @media (min-width: 1700px) {
    height: 65vh;
  }
  @media (min-width: 2000px) {
    height: 75vh;
  }
  &::-webkit-scrollbar {
    width: 10px;
    background-color: #f2f2f2;
  } 
  &::-webkit-scrollbar-thumb{
    background-color: #8f9ea8;
    border-radius: 100px;
  }
`

interface GridProps<T> {
  items: Array<T>
  renderItem: (item: T) => React.ReactNode
}

/**
 * Renders components to a nice grid
 * 
 * Reusable with other types of data
 * 
 * @param items Array of items
 * @param renderItem Function to render each item, automatically assigns types
 * 
 * Usage: <Grid items={youritems} renderItem={<YourComponent />}>
 * 
 * @returns JSX element 
 */
export const Grid = <T,>({ items, renderItem }: GridProps<T>) => {

  return (
    <GridCSS className="overflow-y-auto mt-4 shadow-lg my-4 border-2 border-gray-100 rounded-xl p-2">
      {items.map((repo: T) => (
        renderItem(repo)
      ))}
    </GridCSS>
  )
}