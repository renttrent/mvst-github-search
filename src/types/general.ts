export interface Repo {
  allow_forking: boolean
  archive_url: string
  archived: boolean
  assignees_url: string
  blobs_url: string
  branches_url: string
  clone_url: string
  collaborators_url: string
  comments_url: string
  commits_url: string
  compare_url: string
  contents_url: string
  contributors_url: string
  created_at: string
  default_branch: string
  deployments_url: string
  description: null
  disabled: boolean
  downloads_url: string
  events_url: string
  fork: boolean
  forks: number
  forks_count: number
  forks_url: string
  full_name: string
  git_commits_url: string
  git_refs_url: string
  git_tags_url: string
  git_url: string
  has_downloads: boolean
  has_issues: boolean
  has_pages: boolean
  has_projects: boolean
  has_wiki: boolean
  homepage: string | null
  hooks_url: string
  html_url: string
  id: number
  is_template: boolean
  issue_comment_url: string
  issue_events_url: string
  issues_url: string
  keys_url: string
  labels_url: string
  language: string
  languages_url: string
  license: License | null
  merges_url: string
  milestones_url: string
  mirror_url: string | null
  name: string
  node_id: string
  notifications_url: string
  open_issues: number
  open_issues_count: number
  owner: GithubUser,
  private: boolean
  pulls_url: string
  pushed_at: string
  releases_url: string
  size: number
  ssh_url: string
  stargazers_count: number
  stargazers_url: string
  statuses_url: string
  subscribers_url: string
  subscription_url: string
  svn_url: string
  tags_url: string
  teams_url: string
  topics: Array<string | any>
  trees_url: string
  updated_at: string
  url: string
  visibility: string
  watchers: number
  watchers_count: number
}

interface GithubUser {
  avatar_url: string
  events_url: string
  followers_url: string
  following_url: string
  gists_url: string
  gravatar_id: string
  html_url: string
  id: number
  login: string
  node_id: string
  organizations_url: string
  received_events_url: string
  repos_url: string
  site_admin: boolean
  starred_url: string
  subscriptions_url: string
  type: string
  url: string
}

export interface License {
  key: string
  name: string
  node_id: string
  spdx_id: string
  url: string
}


export interface User {
  avatar_url: string | null
  bio: string | null
  blog: string | null
  collaborators: number | null
  company: string | null
  created_at: string | null
  disk_usage: number | null 
  email: string | null 
  events_url: string | null 
  followers: number 
  followers_url: string | null 
  following: number 
  following_url: string | null 
  gists_url: string | null 
  gravatar_id: string | null 
  hireable: boolean 
  html_url: string 
  id: number 
  location: string | null
  login: string 
  name: string | null
  node_id: string
  oauth: { access_token: string, token_type: string }
  organizations_url: string| null 
  owned_private_repos: number 
  plan: { name: string, space: number, collaborators: number, private_repos: number}
  private_gists: number 
  public_gists: number 
  public_repos: number
  received_events_url: string | null 
  repos_url: string | null 
  site_admin: string | null 
  starred_url: string | null 
  subscriptions_url: string | null 
  total_private_repos: number 
  twitter_username: string | null
  two_factor_authentication: boolean 
  type: string
  updated_at: string 
  url: string 
}