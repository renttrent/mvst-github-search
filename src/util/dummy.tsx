export const getRepo = () => {
  return {
    name: "Github Repo",
    description: "Testing Card for MVST, you should click the MIT license",
    forks: 100,
    stargazers_count: 200,
    language: "Typescript",
    license: {
      key: "MIT",
      name: "MIT",
      node_id: "asdasdagsdf",
      spdx_id: "MIT",
      url: "https://youtu.be/oHg5SJYRHA0"
    },
    updated_at: "2022-06-11T20:49:50.084Z",
    url: "https://youtu.be/oHg5SJYRHA0"
  }
}

export const getRepoList = () => {
  return [
    {
      name: "Github Repo",
      description: "Testing Card for MVST, you should click the MIT license",
      forks: 100,
      stargazers_count: 200,
      language: "Typescript",
      license: {
        key: "MIT",
        name: "MIT",
        node_id: "asdasdagsdf",
        spdx_id: "MIT",
        url: "https://youtu.be/oHg5SJYRHA0"
      },
      updated_at: "2022-06-11T20:49:50.084Z",
      url: "https://youtu.be/oHg5SJYRHA0"
    },
    {
      name: "Github Repo",
      description: "Testing Card for MVST, you should click the MIT license",
      forks: 100,
      stargazers_count: 200,
      language: "Typescript",
      license: {
        key: "MIT",
        name: "MIT",
        node_id: "asdasdagsdf",
        spdx_id: "MIT",
        url: "https://youtu.be/oHg5SJYRHA0"
      },
      updated_at: "2022-06-11T20:49:50.084Z",
      url: "https://youtu.be/oHg5SJYRHA0"
    },
    {
      name: "Github Repo",
      description: "Testing Card for MVST, you should click the MIT license",
      forks: 100,
      stargazers_count: 200,
      language: "Typescript",
      license: {
        key: "MIT",
        name: "MIT",
        node_id: "asdasdagsdf",
        spdx_id: "MIT",
        url: "https://youtu.be/oHg5SJYRHA0"
      },
      updated_at: "2022-06-11T20:49:50.084Z",
      url: "https://youtu.be/oHg5SJYRHA0"
    },
    {
      name: "Github Repo",
      description: "Testing Card for MVST, you should click the MIT license",
      forks: 100,
      stargazers_count: 200,
      language: "Typescript",
      license: {
        key: "MIT",
        name: "MIT",
        node_id: "asdasdagsdf",
        spdx_id: "MIT",
        url: "https://youtu.be/oHg5SJYRHA0"
      },
      updated_at: "2022-06-11T20:49:50.084Z",
      url: "https://youtu.be/oHg5SJYRHA0"
    },
    {
      name: "Github Repo",
      description: "Testing Card for MVST, you should click the MIT license",
      forks: 100,
      stargazers_count: 200,
      language: "Typescript",
      license: {
        key: "MIT",
        name: "MIT",
        node_id: "asdasdagsdf",
        spdx_id: "MIT",
        url: "https://youtu.be/oHg5SJYRHA0"
      },
      updated_at: "2022-06-11T20:49:50.084Z",
      url: "https://youtu.be/oHg5SJYRHA0"
    }
  ]
}