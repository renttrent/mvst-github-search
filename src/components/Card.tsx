import { FaStar } from "react-icons/fa";
import { BiGitRepoForked } from "react-icons/bi";
import { License } from "../types/general";

interface CardProps {
  name: string;
  description: string | null;
  license: License | null;
  stargazers_count: number;
  forks: number;
  language: string;
  updated_at: string;
  html_url: string;
}

/**
 *
 * Card props based on Repo interface.
 *
 * For more see types/general.ts.
 * @param name repo name | string
 * @param description repo description | string
 * @param license repo license link | License
 * @param stargazers_count repo stars nr | number
 * @param forks repo forks nr | number
 * @param language repo most used language | string
 * @param updated_at repo last updated date | string
 *
 * Usage: spread to props object of type Repo
 * @returns JSX element
 */

export const Card = ({
  name: title,
  description: desc,
  license,
  stargazers_count: stars,
  forks,
  language,
  updated_at: lastUpdated,
  html_url,
}: CardProps) => {
  const beautifyDate = (date: string) => {
    const rv = new Date(date);
    return rv.toLocaleDateString("en-US", {
      day: "numeric",
      year: "numeric",
      month: "long",
    });
  };

  return (
    <div
      className="border-2 border-neutral-400 p-3 rounded-md m-3 relative hover:bg-slate-100 cursor-pointer"
      onClick={() => (window.location.href = html_url)}
    >
      <div className="font-bold text-xl">{title}</div>
      <div className="text-lg">{desc}</div>
      <div className="flex flex-row gap-4 items-center">
        <div className="italic text-md text-neutral-600">
          {beautifyDate(lastUpdated)}
        </div>
        {/* @ts-ignore */}
        <div className="text-green-500 font-semibold">{language}</div>
      </div>
      <div className="flex flex-row gap-2 items-center">
        <div className="flex flex-row gap-1 items-center">
          <FaStar className="text-yellow-400" /> {stars}
        </div>
        <div className="flex flex-row gap-1 items-center">
          <BiGitRepoForked className="text-blue-700" /> {forks}
        </div>
        {license && (
          <a href={license.url} className="absolute right-4">
            License - {license.spdx_id}
          </a>
        )}
      </div>
    </div>
  );
};
