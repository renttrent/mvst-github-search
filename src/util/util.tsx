import axios from "axios"
import { Repo, User } from "../types/general"

export const getUserRepos = (user: User) => {
  return axios.get(`https://api.github.com/search/repositories?q=user:${user.login}`, {
    headers: {
      "Authorization": `token ${user.oauth.access_token}`
    }
  })
  .then((res) => {
    const rv = res.data.items.sort(
      (prev: Repo, current: Repo) => {
        //@ts-ignore
        return new Date(current.updated_at) - new Date(prev.updated_at)
      }
    )
    return rv
  })
  .catch((err) => {
    return `Error! ${err.message}`
  })
} 

export const getPublicRepo = (url: string) => {
  return axios.get(url)
  .then((res) => {
    const rv = res.data.sort(
      (prev: Repo, current: Repo) => {
        //@ts-ignore
        return new Date(current.updated_at) - new Date(prev.updated_at)
      }
    )
    return rv
  })
  .catch((err) => {
    return `Error! ${err.message}`
  })
} 

