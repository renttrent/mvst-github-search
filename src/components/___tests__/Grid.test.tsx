import renderer from "react-test-renderer";
import { Repo } from "../../types/general";
import { getRepoList } from "../../util/dummy";
import { Card } from "../Card";
import { Grid } from "../Grid";

const props = {
  name: "test",
  description: "test",
  forks: 2,
  html_url: "test",
  language: "js",
  stargazers_count: 4,
  license: {
    key: "2",
    name: "test",
    node_id: "1sgsfhg",
    spdx_id: "MIT",
    url: "test",
  },
  updated_at: "testdate",
};

it("Should render normally", () => {
  const grid = renderer
    .create(
      <Grid
        //@ts-ignore Ignoring since dummy repo list does not have correct types.
        items={[props]}
        renderItem={() => <Card key={2} {...props} />}
      />
    )
    .toJSON();
  expect(grid).toMatchSnapshot();
});

it("Should render with other component", () => {
  const grid = renderer
    .create(
      <Grid
        items={[{ name: "rei", degree: "cs", id: 1 }]}
        renderItem={(person) => (
          <div key={person.id}>
            <div>{person.id}</div>
            <div>{person.name}</div>
            <div>{person.degree}</div>
          </div>
        )}
      />
    )
    .toJSON();
  expect(grid).toMatchSnapshot();
});
