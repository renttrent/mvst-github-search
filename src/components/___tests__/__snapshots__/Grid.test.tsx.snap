// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render normally 1`] = `
<div
  className="overflow-y-auto mt-4 shadow-lg my-4 border-2 border-gray-100 rounded-xl p-2 css-o4lsx0"
>
  <div
    className="border-2 border-neutral-400 p-3 rounded-md m-3 relative hover:bg-slate-100 cursor-pointer"
    onClick={[Function]}
  >
    <div
      className="font-bold text-xl"
    >
      test
    </div>
    <div
      className="text-lg"
    >
      test
    </div>
    <div
      className="flex flex-row gap-4 items-center"
    >
      <div
        className="italic text-md text-neutral-600"
      >
        Invalid Date
      </div>
      <div
        className="text-green-500 font-semibold"
      >
        js
      </div>
    </div>
    <div
      className="flex flex-row gap-2 items-center"
    >
      <div
        className="flex flex-row gap-1 items-center"
      >
        <svg
          className="text-yellow-400"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
          />
        </svg>
         
        4
      </div>
      <div
        className="flex flex-row gap-1 items-center"
      >
        <svg
          className="text-blue-700"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.559 8.855c.166 1.183.789 3.207 3.087 4.079C11 13.829 11 14.534 11 15v.163c-1.44.434-2.5 1.757-2.5 3.337 0 1.93 1.57 3.5 3.5 3.5s3.5-1.57 3.5-3.5c0-1.58-1.06-2.903-2.5-3.337V15c0-.466 0-1.171 2.354-2.065 2.298-.872 2.921-2.896 3.087-4.079C19.912 8.441 21 7.102 21 5.5 21 3.57 19.43 2 17.5 2S14 3.57 14 5.5c0 1.552 1.022 2.855 2.424 3.313-.146.735-.565 1.791-1.778 2.252-1.192.452-2.053.953-2.646 1.536-.593-.583-1.453-1.084-2.646-1.536-1.213-.461-1.633-1.517-1.778-2.252C8.978 8.355 10 7.052 10 5.5 10 3.57 8.43 2 6.5 2S3 3.57 3 5.5c0 1.602 1.088 2.941 2.559 3.355zM17.5 4c.827 0 1.5.673 1.5 1.5S18.327 7 17.5 7 16 6.327 16 5.5 16.673 4 17.5 4zm-4 14.5c0 .827-.673 1.5-1.5 1.5s-1.5-.673-1.5-1.5.673-1.5 1.5-1.5 1.5.673 1.5 1.5zM6.5 4C7.327 4 8 4.673 8 5.5S7.327 7 6.5 7 5 6.327 5 5.5 5.673 4 6.5 4z"
          />
        </svg>
         
        2
      </div>
      <a
        className="absolute right-4"
        href="test"
      >
        License - 
        MIT
      </a>
    </div>
  </div>
</div>
`;

exports[`Should render with other component 1`] = `
<div
  className="overflow-y-auto mt-4 shadow-lg my-4 border-2 border-gray-100 rounded-xl p-2 css-o4lsx0"
>
  <div>
    <div>
      1
    </div>
    <div>
      rei
    </div>
    <div>
      cs
    </div>
  </div>
</div>
`;
